### PR TITLE
feat(jira): components + Sprint + Structure generators/automation/delete (#197 #198 #179 #180)

### PR DESCRIPTION
**Pull request by andreymaznyak**
Client-level bundle closing four priority Jira feature gaps in a single PR so cloud integration can pick them up as one atomic bump.

Each commit is standalone-reviewable and maps 1:1 to its upstream issue.

| Issue | Commit | Scope |
|-------|--------|-------|
| #197 (priority:high) | `1441efc feat(jira): add components support to create_issue and update_issue` | `CreateIssueInput.components: Vec<String>`, `UpdateIssueInput.components: Option<Vec<String>>`, payload `{ fields: { components: [{id}] } }`, 3 httpmock tests |
| #198 (priority:high) | `7adebd5 feat(jira): Sprint + ...` | Agile REST helper, `Sprint`/`SprintState`/`AssignToSprintInput`, `IssueProvider::get_board_sprints` + `assign_to_sprint`, 3 tests |
| #179 | `7adebd5 (same commit)` | `StructureGenerator` + three trait methods (`get_structure_generators`, `add_structure_generator`, `sync_structure_generator`), 1 lifecycle test |
| #180 | `7adebd5 (same commit)` | `UpdateStructureAutomationInput` + three trait methods (`delete_structure`, `update_structure_automation`, `trigger_structure_automation`), 2 tests |

## What's in

- **Provider trait extensions** — all new methods default to `ProviderUnsupported`, so ClickUp/GitLab/GitHub providers don't need changes.
- **Jira client implementations** — wire to existing `structure_*` helpers for Structure endpoints, new `agile_url`/`agile_get`/`agile_post_void` for Agile REST.
- **Tests** — 9 new httpmock-backed cases covering happy path + edge cases (empty components skipped, `SprintState::All` omits query param, `jira#` prefix stripping on sprint assign).

Verification:
- `cargo test --workspace` ✅ all passing (jira: 155, executor: 183, format-pipeline: 65, rest clean)
- `cargo clippy --workspace --all-targets -- -D warnings` ✅ clean
- `cargo fmt --all --check` ✅ clean

## What's explicitly **not** in (follow-ups)

- **#177 `UserProvider` trait + Slack user_profile** — cross-cutting refactor (new trait, refactor existing Jira `get_users`, new Slack client code). Scoped out so this PR stays reviewable. Will be a dedicated follow-up.
- **Tool registrations in `crates/devboy-executor/src/tools.rs` / dispatch in `executor.rs` / schema enrichment in `jira/src/enricher.rs`** — new tools for `get_board_sprints`, `assign_to_sprint`, `get_structure_generators`, `add_structure_generator`, `sync_structure_generator`, `delete_structure`, `update_structure_automation`, `trigger_structure_automation` will be a separate follow-up so this diff stays focused on the client-level contract. The existing `create_issue`/`update_issue` tools already pick up the new `components` field through schema pass-through.
- **Cloud-side wrappers** (NestJS providers in `apps/devboy-api/.../mcp/tools/`, entries in `MCP_TOOLS_REGISTRY`) — separate cloud MR after a v0.20.0 bump.

Refs: #197 #198 #179 #180.

### CODE REVIEW

File: crates/plugins/api/jira/src/client.rs [hunk 1]
```diff
@@ -2332,6 +2397,150 @@ impl IssueProvider for JiraClient {
...
+                .get("id")
+                .and_then(|v| v.as_str())
+                .unwrap_or_default()
+                .to_string(),
+            generator_type: resp
+                .get("type")
+                .and_then(|v| v.as_str())
+                .unwrap_or_default()
+                .to_string(),
+            spec: resp.get("spec").cloned().unwrap_or(serde_json::Value::Null),
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/jira/src/client.rs]: ✅ Fixed in commit `5868e83` — replaced `unwrap_or_default()` salvage with a typed `Resp { id: String, type, spec }` struct. Missing `id`/`type` now surface as a deserialise error instead of producing a `StructureGenerator` with empty id.

File: crates/plugins/api/jira/src/client.rs [hunk 2]
```diff
@@ -2332,6 +2397,150 @@ impl IssueProvider for JiraClient {
...
+    async fn update_structure_automation(
+        &self,
+        input: devboy_core::UpdateStructureAutomationInput,
+    ) -> Result<()> {
+        let _: serde_json::Value = self
+            .structure_put(
+                &format!("/structure/{}/automation", input.structure_id),
+                &input.config,
+            )
+            .await?;
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/jira/src/client.rs]: ✅ Fixed in commit `5868e83` — Structure plugin exposes both rule-scoped and collection-level endpoints. Added `automation_id: Option<String>` to `UpdateStructureAutomationInput`:
- `Some(id)` → `PUT /structure/{structureId}/automation/{automationId}` (rule-scoped)
- `None` → `PUT /structure/{structureId}/automation` (replace whole set)

Added a dedicated `test_structure_automation_rule_scoped` for the rule-scoped path.

File: crates/devboy-executor/src/executor.rs [hunk 3]
```diff
@@ -719,6 +719,15 @@ async fn execute_create_issue(
...
     let custom_fields = args.get("customFields").cloned();
+    let components: Vec<String> = args
+        .get("components")
+        .and_then(|v| v.as_array())
+        .map(|arr| {
+            arr.iter()
+                .filter_map(|v| v.as_str().map(|s| s.to_string()))
+                .collect()
+        })
+        .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: ✅ Fixed in commit `5868e83` — added typed field `components: Vec<String>` on `CreateIssueParams` (serde-parsed). Non-array / non-string input now errors fast instead of being silently dropped.

File: crates/devboy-executor/src/executor.rs [hunk 4]
```diff
@@ -766,6 +776,16 @@ async fn execute_update_issue(
...
+    // Issue #197 — accept `components: [id, ...]` on update. If the key is
+    // absent we leave them untouched (None); an empty array clears them.
+    let components: Option<Vec<String>> =
+        args.get("components")
+            .and_then(|v| v.as_array())
+            .map(|arr| {
+                arr.iter()
+                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
+                    .collect()
+            });
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-executor/src/executor.rs]: ✅ Fixed in commit `5868e83` — added typed `components: Option<Vec<String>>` on `UpdateIssueParams`. `None` (key absent) leaves untouched, `Some([])` clears, `Some([...])` replaces. Non-array / non-string input fails fast.

File: crates/devboy-core/src/types.rs [hunk 5]
```diff
@@ -166,6 +166,12 @@ pub struct CreateIssueInput {
     /// ClickUp: Array `[{"id": "field_id", "value": val}]` — set via separate API.
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub custom_fields: Option<Value>,
+    /// Component IDs to associate with the issue (Jira-only, issue #197).
+    /// Each entry is a component `id` obtained from project metadata
+    /// (`JiraComponent.id`). Ignored by providers that don't have a
+    /// first-class Components concept (GitHub/GitLab/ClickUp).
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/devboy-core/src/types.rs]: ✅ Fixed in commit `5868e83` — switched contract to **names** end-to-end. `CreateIssueInput.components` / `UpdateIssueInput.components` now explicitly document that the entries are component **names** (from `JiraComponent.name`, matching the enricher's enum values). Jira accepts either id or name in `fields.components`; names keep callers' mental model consistent with the schema enum.

File: crates/plugins/api/jira/src/client.rs [hunk 6]
```diff
@@ -1650,6 +1691,19 @@ impl IssueProvider for JiraClient {
...
+            None
+        } else {
+            Some(
+                input
+                    .components
+                    .into_iter()
+                    .map(|id| crate::types::ComponentRef { id })
+                    .collect(),
+            )
+        };
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/jira/src/client.rs:1691]: ✅ Fixed in commit `5868e83` — `ComponentRef` now serialises `{ "name": "..." }`. End-to-end path is names: enricher enum → tool input → client payload. Updated existing `test_create_issue_with_components` / `test_update_issue_replaces_components` to check `\"components\":[{\"name\":\"Backend\"}]` shape.

File: crates/plugins/api/jira/src/types.rs [hunk 7]
```diff
@@ -284,6 +284,17 @@ pub struct CreateIssueFields {
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub components: Option<Vec<ComponentRef>>,
+}
+
+/// Component reference used in create/update issue payloads (issue #197).
+/// Jira expects `{ "id": "10001" }` — name-based references also work in
+/// Cloud, but id is more stable across rename.
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct ComponentRef {
+    pub id: String,
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/jira/src/types.rs]: ✅ Fixed in commit `5868e83` — `ComponentRef { name: String }`. Docstring updated to explain the decision (Jira accepts both id and name; name lines up with the enricher enum). See sibling reply on `CreateIssueInput.components` for the end-to-end rationale.

File: crates/plugins/api/jira/src/client.rs [hunk 8]
```diff
@@ -576,6 +576,47 @@ impl JiraClient {
...
+            .map_err(|e| Error::Http(e.to_string()))?;
+        let status = response.status();
+        if !status.is_success() {
+            let (content_type, body) = read_structure_error_body(response).await;
+            return Err(structure_error_from_status(
+                status.as_u16(),
+                &content_type,
+                body,
+            ));
+        }
```
[the remark this answers is not in the document]
Comment (andreymaznyak) [reply crates/plugins/api/jira/src/client.rs]: ✅ Fixed in commit `5868e83` — `agile_get` now uses the generic `self.get`, `agile_post_void` uses `self.post_no_content`. Both go through `handle_response` which doesn't carry the Structure-plugin hint. Agile 404 / SSO redirects are no longer misattributed to a missing Structure plugin.

**Comment by andreymaznyak** [comment]
## 🔍 All review comments addressed

### Copilot (8 inline findings)

| # | Finding | Commit |
|---|---------|--------|
| 1 | `add_structure_generator` `unwrap_or_default()` → typed Resp | `5868e83` |
| 2 | `update_structure_automation` endpoint — collection vs rule-scoped | `5868e83` (new `automation_id: Option<String>`) |
| 3 | `components` executor parsing → serde on `CreateIssueParams` | `5868e83` |
| 4 | `components` executor parsing → serde on `UpdateIssueParams` | `5868e83` |
| 5 | `components` docstring: ID vs name contract | `5868e83` (switched to names end-to-end) |
| 6 | `ComponentRef` payload shape ID vs name mismatch | `5868e83` (`{name: String}`) |
| 7 | `ComponentRef` doc: name-based throughout | `5868e83` |
| 8 | Agile helpers leaking Structure plugin hint | `5868e83` (use generic `handle_response`) |

Plus reply threads under each inline thread with concrete diff.

### Codex review (3 findings)

| Severity | Finding | Commit |
|----------|---------|--------|
| P1 | component **name** vs ID (chicken-and-egg — callers can't know ids) | `5868e83` — duplicated of Copilot #5/#6/#7, **confirmed by both sources** |
| P2 | `components` schema property missing in `create_issue`/`update_issue` input_schema | `0e14816` (new `s.add_property("components", ...)`) |
| P2 | `/board/{id}/sprint` pagination not walked | `0e14816` (`startAt` + `isLast` loop, 5k cap, new pagination test) |

### Both sources independently converged on the "name vs id" issue 

— strong signal, now fixed end-to-end: enricher enum (name) → tool input (name) → `ComponentRef { name: String }` → Jira payload.

All workspace tests green, clippy clean, fmt clean after both fix commits. Ready for re-review.
